src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -208,6 +206,62 @@ protected ExecutionStrategy(DataFetcherExceptionHandler dataFetcherExceptionHand
+        ValueFetcher valueFetcher = new ValueFetcher(executionContext, false) {
+            InstrumentationContext<Object> fetchCtx;
+
+            @Override
+            protected FetchedValue handleExceptionWhileFetching(Field field, ExecutionPath path, Throwable exception, DataFetchingEnvironment environment, Map<String, Object> argumentValues, GraphQLFieldDefinition fieldDefinition, ExecutionPath executionPath) {
+                DataFetcherExceptionHandlerParameters handlerParameters = DataFetcherExceptionHandlerParameters.newExceptionParameters()
+                        .executionContext(executionContext)
+                        .dataFetchingEnvironment(environment)
+                        .argumentValues(argumentValues)
+                        .field(field)
+                        .fieldDefinition(fieldDefinition)
+                        .path(parameters.getPath())
+                        .exception(exception)
+                        .build();
+                dataFetcherExceptionHandler.accept(handlerParameters);
+                parameters.deferredErrorSupport().onFetchingException(parameters, exception);
+                return new FetchedValue(null, null);
+            }
+
+            @Override
+            protected DataFetcher beforeCallingDataFetcher(DataFetcher dataFetcher,
+                                                           ExecutionContext executionContext,
+                                                           GraphQLFieldDefinition fieldDef,
+                                                           DataFetchingEnvironment environment) {
+                Instrumentation instrumentation = executionContext.getInstrumentation();
+
+                InstrumentationFieldFetchParameters instrumentationFieldFetchParams = new InstrumentationFieldFetchParameters(executionContext,
+                        fieldDef,
+                        environment,
+                        parameters,
+                        dataFetcher instanceof TrivialDataFetcher);
+                fetchCtx = instrumentation.beginFieldFetch(instrumentationFieldFetchParams);
+                return instrumentation.instrumentDataFetcher(dataFetcher, instrumentationFieldFetchParams);
+            }
+
+            @Override
+            protected void afterCallingDataFetcher(CompletableFuture<Object> dataFetcherCallCF) {
+                fetchCtx.onDispatched(dataFetcherCallCF);
+            }
+
+            @Override
+            protected void handleResult(Object value, Throwable throwable) {
+                fetchCtx.onCompleted(value, throwable);
+            }
+        };
+        ExecutionStepInfoFactory executionStepInfoFactory = new ExecutionStepInfoFactory(executionContext);
+        ExecutionStepInfo subStepInfo = executionStepInfoFactory.newExecutionStepInfoForSubField(parameters.getField(), parameters.getExecutionStepInfo());
+        return valueFetcher
```

Comment:
I still think `ExecutionStepInfoFactory` should be a singleton that takes the per request things (execitionCOntext et al) as parameters